gt/main.c

```diff
@@ -789,22 +944,38 @@ gt_proc(void *arg)
 			 */
 			ret = gt_parse_incoming_pkt(m, &pkt_info);
 			if (ret < 0) {
-				switch (pkt_info.outer_ethertype) {
-				case ETHER_TYPE_IPv4:
-					add_pkt_acl(&acl4, m);
-					continue;
-				case ETHER_TYPE_IPv6:
-					add_pkt_acl(&acl6, m);
+				gt_process_unparsed_incoming_pkt(
+					&acl4, &acl6, &num_arp, arp_bufs,
+					m, pkt_info.outer_ethertype);
+				continue;
+			}
+
+			/*
+			 * If it is a fragmented packet,
+			 * then try to reassemble.
+			 */
+			if (pkt_info.frag) {
+				m = gt_reassemble_incoming_pkt(
+					m, cur_tsc, &pkt_info,
+					&death_row, instance);
```

Comment:
Consider a deployment in which fragmented packets are not common, that is, it only happens occasionally every hour. If the segment of a packet is lost, it'll take about 30min (1/2h assuming uniform distribution) for another segmented packet to show up and release the memory of the expired segments. Then, the punishment will be **too** late. That is, we need a timer to search for expired segments when no fragments arrive. Whenever a segment arrives, the timer should be reset.

Reply:
If the segmented packets are not common, will they cause critical issues to the request channel? Moreover, the punishment policy may expire before the next one arrives even we try to punish them. It seems that we may have to iterate over the whole fragmentation table to search the expired ones.

Reply:
The example I constructed is not about the request channel, but the fact that we are not punishing incomplete fragmented packets **when** they should, and keeping fragments in memory for way pass the time we could've freed them.

Reply:
I agree. We need to iterate over the whole fragmentation table to search the expired fragments.

Please ignore my suggestion to reset the timer when a segment arrives because, as you stated in another comment, neither `rte_ipv4_frag_reassemble_packet()`, nor `rte_ipv6_frag_reassemble_packet()` scans the whole table of fragments.

To avoid scanning the whole table of fragments all at once, what would make GT blocks to stop processing packets for "too much" time, each time the timer runs, we inspect a single bucket. Of course, we'll need a variable in between calls to know which bucket is next.

Reply:
I see your point. I think the expired fragments will only affect the memory usage, because more fragments in the table may lead to more buffers for them. If we would like to trade off the memory usage and scan time overhead, I can add a new patch which extends the DPDK Library to handle such scenarios.

Reply:
To be clear here, there are two problems here: 1. the memory usage that you just pointed, 2. expired segments that need to be punished. Both these problems require that we scan the whole table of segments once in a while with a timer. If we had only the first issue, we could ignore the scanning and take the extra memory as a reasonable tradeoff.

Reply:
That makes sense. Thanks